### PR TITLE
Auto generate Triggers OIDC identity service account and expose in its status

### LINKED ISSUE #3521: [control-plane]: Support auto generation of Triggers identity service account and expose in `AuthStatus`
**Problem** 

As the [Eventing OIDC feature track](https://docs.google.com/document/d/1e7UgNTkL0Br5Da09Rg2ieVmhKJo4VXuBj-mHT9NCujY/edit#bookmark=id.gvbj3362t2fy) describes, the `AuthStatus` is meant to provide the generated service account name in the resource status.

When https://github.com/knative/eventing/issues/7173 is done, we should:
* when the feature flag (see https://github.com/knative/eventing/issues/7174) is disabled:
  * no change in behavior
* when the feature flag (see https://github.com/knative/eventing/issues/7174) is enabled:
  * create the service account for the Triggers identity
  * expose the service accounts name in `.status.auth.serviceAccountName`

**Additional Information:**
* See implementation for trigger in eventing-core: https://github.com/knative/eventing/issues/7222

### PR DESCRIPTION
**Pull request by xiangpingjiang**
Support auto generation of Triggers identity service account and expose in AuthStatus

Fixes #3521 

## Proposed Changes

-
-
-

**Release Note**

```release-note
Support auto generation of Triggers identity service account and expose in AuthStatus
```

**Docs**

### CODE REVIEW

**Comment by creydr** [comment]
Hi @xiangpingjiang,
thanks a lot for your PR and working on this.
Can you add a test for this in `control-plane/pkg/reconciler/trigger/trigger_test.go`?

/ok-to-test

File: control-plane/pkg/reconciler/trigger/trigger.go [hunk 1]
```diff
@@ -61,10 +63,11 @@ type Reconciler struct {
 	*base.Reconciler
 	*FlagsHolder
 
-	BrokerLister    eventinglisters.BrokerLister
-	ConfigMapLister corelisters.ConfigMapLister
-	EventingClient  eventingclientset.Interface
-	Resolver        *resolver.URIResolver
+	BrokerLister         eventinglisters.BrokerLister
+	ConfigMapLister      corelisters.ConfigMapLister
+	serviceAccountLister corelisters.ServiceAccountLister
```
Comment (Leo6Leo) [inline control-plane/pkg/reconciler/trigger/trigger.go:68]: Do we want to make this become public by renaming it to `ServiceAccountLister`?

Comment (Leo6Leo) [reply control-plane/pkg/reconciler/trigger/trigger.go:68]: As I see for [namespaced broker,](https://github.com/knative-extensions/eventing-kafka-broker/blob/9541bb6f66b643c470291ca2753fee7c1fb6ad5e/control-plane/pkg/reconciler/broker/namespaced_broker.go#L71) it is public.

File: control-plane/pkg/reconciler/trigger/trigger_test.go [hunk 2]
```diff
@@ -123,6 +125,180 @@ func TestTriggerWithNewFiltersReconciler(t *testing.T) {
...
+	eventing.RegisterAlternateBrokerConditionSet(base.IngressConditionSet)
+
+	t.Parallel()
+
+	for _, f := range Formats {
+		triggerReconciliationWithOIDC(t, f, *DefaultEnv)
+	}
+}
+
+func triggerReconciliationWithOIDC(t *testing.T, format string, env config.Env) {
```
Comment (creydr) [inline control-plane/pkg/reconciler/trigger/trigger_test.go:138]: I think we can integrate them into the `triggerReconcilation()` function.
You can use the following to set the feature flags for a test:
```
Name: "OIDC: ...",
Ctx: feature.ToContext(context.Background(), feature.Flags{
	feature.OIDCAuthentication: feature.Enabled,
}),
Objects: ...
```

That way we align more with the existing

Comment (xiangpingjiang) [reply control-plane/pkg/reconciler/trigger/trigger_test.go:138]: > I think we can integrate them into the `triggerReconcilation()` function. You can use the following to set the feature flags for a test:
> 
> ```
> Name: "OIDC: ...",
> Ctx: feature.ToContext(context.Background(), feature.Flags{
> 	feature.OIDCAuthentication: feature.Enabled,
> }),
> Objects: ...
> ```
> 
> That way we align more with the existing

I have tried that, but it didnt work .

Comment (creydr) [reply control-plane/pkg/reconciler/trigger/trigger_test.go]: @xiangpingjiang: https://github.com/knative-extensions/eventing-kafka-broker/pull/3612 merged. When you rebase, you should be able to use it in `triggerReconcilation()` with a configured `Ctx`.

**Review by creydr — CHANGES_REQUESTED** [review]
Hello @xiangpingjiang,
thanks for your PR. When I tested it, I didn't see a service account created for a trigger and the kafka-controller was crashing. Can you check on it?
And we should add some logic to reconcile triggers, if the features config map gets updated. e.g. for the following:
* disabled -> enabled: create sa and set .status.auth
* enabled -> disabled: clear .status.auth (depending on how fast https://github.com/knative/eventing/pull/7570 gets in, you can also add/reuse the logic from there to remove the service account again)

**Comment by xiangpingjiang** [comment]
> Hello @xiangpingjiang, thanks for your PR. When I tested it, I didn't see a service account created for a trigger and the kafka-controller was crashing. Can you check on it? And we should add some logic to reconcile triggers, if the features config map gets updated. e.g. for the following:
> 
> * disabled -> enabled: create sa and set .status.auth
> * enabled -> disabled: clear .status.auth (depending on how fast [Remove OIDC service account, when OIDC feature is disabled again knative/eventing#7570](https://github.com/knative/eventing/pull/7570) gets in, you can also add/reuse the logic from there to remove the service account again)

Hello, @creydr 
Where I can find a simple example to show how to use the eventing-kafka-broker?
I have read the doc https://knative.dev/docs/eventing/brokers/broker-types/kafka-broker/  , but didn't find it.

**Comment by creydr** [comment]
Hi @xiangpingjiang,

https://github.com/knative-extensions/eventing-kafka-broker/blob/main/DEVELOPMENT.md#installation provides useful information how to install EKB (and general developement information).

After the cluster is setup with EKB, you can create a broker & trigger the usual way. e.g. via 
```
apiVersion: eventing.knative.dev/v1
kind: Broker
metadata:
  annotations:
    eventing.knative.dev/broker.class: Kafka
  name: my-kafka-broker
```
or via the `kn` client with `kn broker create my-kafka-broker --class Kafka`.

If the docs are not clear or you miss anything, feel free to update them or create an issue with what's missing so we can improve it.

**Comment by creydr** [comment]
> hello @creydr We have a quickstart for eventing https://knative.dev/docs/getting-started/first-source/ Do we have something similar for using EKB?

Hey @xiangpingjiang,
the quickstart for eventing is about how to develop a source in general. Not sure what you have in mind for EKB here, but probably something to bring up in a Workgroup meeting?

**Comment by xiangpingjiang** [comment]
> > hello @creydr We have a quickstart for eventing https://knative.dev/docs/getting-started/first-source/ Do we have something similar for using EKB?
> 
> Hey @xiangpingjiang, the quickstart for eventing is about how to develop a source in general. Not sure what you have in mind for EKB here, but probably something to bring up in a Workgroup meeting?

@creydr 
For example.  After installing kafka-broker https://knative.dev/docs/eventing/brokers/broker-types/kafka-broker/#installation  ,  how to create a broker and a trigger , and set the source and sink to run a simple demo

**Comment by creydr** [comment]
Hi @xiangpingjiang,
do you need any more info to proceed with this?

**Comment by xiangpingjiang** [comment]
> Hi @xiangpingjiang, do you need any more info to proceed with this?

Hello @creydr 
It's ok for now. Thank you

**Review by creydr — CHANGES_REQUESTED** [review]
Hey @xiangpingjiang,
thanks for your work so far. Seems the current reconciler e2e test failures are due to a nil pointer reference in the namespaced broker.
You need to setup the serviceAccountLister in the NamespacedReconciler too. Something like
```
	serviceaccountInformer := serviceaccountinformer.Get(ctx)

	reconciler := &NamespacedReconciler{
		Reconciler: &base.Reconciler{
				...
		},
		...
		ServiceAccountLister:       serviceaccountInformer.Lister(),
	}
```
in namespaced_controller.go (`NewNamespacedController`).
And probably some other changes (similar what you are doing in controller.go (reconcile triggers when OIDC SA changes, reconcile triggers when features configmap changes))

File: control-plane/pkg/reconciler/trigger/controller.go
[the remark this answers is not in the document]
Comment (xiangpingjiang) [reply control-plane/pkg/reconciler/trigger/controller.go:1]: Hello @creydr 
I didn't really understand why this block code can filter the features configmap?

[the remark this answers is not in the document]
Comment (creydr) [reply control-plane/pkg/reconciler/trigger/controller.go:1]: Hey @xiangpingjiang,
the `feature.NewStore` is a helper around an `UntypedStore`, but already configure for the feature configmap (see [here](https://github.com/knative/eventing/blob/54f3952a38e7ff998a7c71a050be62b2915a3de0/pkg/apis/feature/store.go#L28)). An  [`UntypedStore`](https://github.com/knative/pkg/blob/b6659cc450661dcd0d805525cf1e3e859015ca2d/configmap/store.go#L50) allows you to handle configmaps as a simple "store" and register watchers on it. I hope this helps to understand it

**Comment by creydr** [comment]
seems to be a flake
```
    steps.go:47: sources.knative.dev/v1beta1, Resource=kafkasources did not become ready, timed out waiting for the condition
--- FAIL: TestKnativeKafkaCloudEventExtensionsAdded/KnativeKafkaCEExtensionsAdded/sourceAddsKnativeKafkaCEExtensions/Setup/kafka_source_is_ready (127.37s)
```

/retest-required

**Comment by creydr** [comment]
```
--- FAIL: TestTriggerNoFinalizer/0/triggerNoFinalizerOnBrokerNotFound/Assert/eventually_trigger_has_no_finalizer (20.03s)
```
Flake AFAIK

/retest-required

**Review by creydr — APPROVED** [review]
Looks good so far from my side!
Thanks a lot for your work on this @xiangpingjiang 

/lgtm

**Comment by creydr** [comment]
```
FAIL test/e2e_new.TestTriggerNoFinalizer/0 (814.26s)
```

/retest

**Pull request merged**